Phoenix/hello/lib/hello/shopping_cart.ex

```diff
@@ -0,0 +1,270 @@
+defmodule Hello.ShoppingCart do
+  @moduledoc """
+  The ShoppingCart context.
+  """
+
+  import Ecto.Query, warn: false
+  alias Hello.Repo
+  alias Hello.Catalog
+  alias Hello.ShoppingCart.{Cart, CartItem}
+
+  def prune_cart_items(%Cart{} = cart) do
+    {_, _} = Repo.delete_all(from(i in CartItem, where: i.cart_id == ^cart.id))
+    {:ok, reload_cart(cart)}
+  end
+
+  def get_cart_by_user_uuid(user_uuid) do
+    Repo.one(
+      from(c in Cart,
+        where: c.user_uuid == ^user_uuid,
+        left_join: i in assoc(c, :items),
+        left_join: p in assoc(i, :product),
+        order_by: [asc: i.inserted_at],
+        preload: [items: {i, product: p}]
+      )
+    )
+  end
+
+  def total_item_price(%CartItem{} = item) do
+    Decimal.mult(item.product.price, item.quantity)
+  end
+
+  def total_cart_price(%Cart{} = cart) do
+    Enum.reduce(cart.items, 0, fn item, acc ->
+      item
+      |> total_item_price()
+      |> Decimal.add(acc)
+    end)
+  end
+
+  @doc """
+  Returns the list of carts.
+
+  ## Examples
+
+      iex> list_carts()
+      [%Cart{}, ...]
+
+  """
+  def list_carts do
+    Repo.all(Cart)
+  end
+
+  @doc """
+  Gets a single cart.
+
+  Raises `Ecto.NoResultsError` if the Cart does not exist.
+
+  ## Examples
+
+      iex> get_cart!(123)
+      %Cart{}
+
+      iex> get_cart!(456)
+      ** (Ecto.NoResultsError)
+
+  """
+  def get_cart!(id), do: Repo.get!(Cart, id)
+
+  @doc """
+  Creates a cart.
+
+  ## Examples
+
+      iex> create_cart(%{field: value})
+      {:ok, %Cart{}}
+
+      iex> create_cart(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def create_cart(user_uuid) do
+    %Cart{user_uuid: user_uuid}
+    |> Cart.changeset(%{})
+    |> Repo.insert()
+    |> case do
+      {:ok, cart} -> {:ok, reload_cart(cart)}
+      {:error, changeset} -> {:error, changeset}
+    end
+  end
+
+  defp reload_cart(%Cart{} = cart), do: get_cart_by_user_uuid(cart.user_uuid)
+
+  def add_item_to_cart(%Cart{} = cart, %Catalog.Product{} = product) do
+    %CartItem{quantity: 1, price_when_carted: product.price}
+    |> CartItem.changeset(%{})
+    |> Ecto.Changeset.put_assoc(:cart, cart)
```

Comment:
@samu975 what is an association?

Reply:
An association is used to associate data with other data. There are three types of associations. has_many, to say that data belongs to other data or has several of some data. belong_to, to associate one data is related with another. has_one, to say that x data type has a data type.